tapdb/universe_federation.go

```diff
@@ -174,4 +211,142 @@ func (u *UniverseFederationDB) LogNewSyncs(ctx context.Context,
 	})
 }
 
-var _ universe.FederationLog = (*UniverseFederationDB)(nil)
+// UpsertFederationSyncConfig upserts both the general and universe specific
+// federation sync configs.
+func (u *UniverseFederationDB) UpsertFederationSyncConfig(
+	ctx context.Context, generalSyncConfig *universe.FedGeneralSyncConfig,
+	uniSyncConfigs []*universe.FedUniSyncConfig) error {
+
+	var writeTx UniverseFederationOptions
+	return u.db.ExecTx(ctx, &writeTx, func(db UniverseServerStore) error {
+		// Upsert general federation sync config.
+		err := db.SetFederationGeneralSyncConfig(ctx,
+			generalSyncConfig.ProofTypes.String())
+		if err != nil {
+			return err
+		}
+
+		// Upsert universe specific sync configs.
+		for i := range uniSyncConfigs {
+			var (
+				config = uniSyncConfigs[i]
+
+				uniID        = config.UniverseID
+				assetIDBytes = uniID.AssetID[:]
+			)
+
+			// If the group key is set, then we'll serialize it
+			// into bytes.
+			var pubKey []byte
+			if uniID.GroupKey != nil {
+				pubKey = uniID.GroupKey.SerializeCompressed()
+			}
+
+			params := UpsertFedUniSyncConfigParams{
+				AssetID:   assetIDBytes,
+				GroupKey:  pubKey,
+				ProofType: config.ProofTypes.String(),
+			}
+			err := db.UpsertFederationUniSyncConfig(ctx, params)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+// QueryFederationSyncConfigs returns the general and universe specific
+// federation sync configs.
+func (u *UniverseFederationDB) QueryFederationSyncConfigs(
+	ctx context.Context) (*universe.FedGeneralSyncConfig,
+	[]*universe.FedUniSyncConfig, error) {
+
+	var (
+		readTx UniverseFederationOptions
+
+		generalConfig *universe.FedGeneralSyncConfig
+		uniConfigs    []*universe.FedUniSyncConfig
+	)
+
+	err := u.db.ExecTx(ctx, &readTx, func(db UniverseServerStore) error {
+		// Query for general sync config.
+		proofTypesStr, err := db.QueryFederationGeneralSyncConfig(
+			ctx,
+		)
+		if err != nil {
+			return err
+		}
+		proofTypes, err := universe.NewSyncProofTypesFromString(
+			proofTypesStr,
+		)
+		if err != nil {
+			return err
+		}
+
+		generalConfig = &universe.FedGeneralSyncConfig{
+			ProofTypes: proofTypes,
+		}
+
+		// If no rows in result set, then we haven't set the config
+		// yet, so we'll default to true.
+		if errors.Is(err, sql.ErrNoRows) {
```

Comment:
I think you want this check above? Since we use the value there without checking the error.

Reply:
Bumping this so not forgotten, or maybe needs a slight rework to the control flow?